extras/foreman_protector/dnf/foreman-protector.py

```diff
@@ -0,0 +1,76 @@
+import dnf
+import dnf.exceptions
+from dnfpluginscore import _, logger
+
+import configparser
+package_whitelist = set()
+
+class ForemanProtector(dnf.Plugin):
+    name = 'foreman-protector'
+    config_name = 'foreman-protector'
+
+    def __init__(self,base,cli):
+        self.base = base
+        self.cli = cli
+
+    def config(self):
+        global fileurl
+
+        try:
+             parser = self.read_config(self.base.conf)
+        except Exception as e:
+            raise dnf.exceptions.Error(_("Parsing file failed: {}").format(str(e)))
+
+        ## Need to fix the if condition
+        if parser.has_section('main'):
+            fileurl = parser.get('main', 'whitelist')
+        else:
+            raise dnf.exceptions.Error(_('Incorrect plugin configuration!'))
+
+    def _load_whitelist(self):
+        global package_whitelist
+        try:
+            if fileurl:
+                llfile = open(fileurl, 'r')
+                for line in llfile.readlines():
+                    if line.startswith('#') or line.strip() == '':
+                        continue
+
+                    package_whitelist.add(line.rstrip().lower())
+                llfile.close()
+        except urlgrabber.grabber.URLGrabError as e:
+            raise dnf.exceptions('Unable to read Foreman protector"s configuration: %s' % e)
+
+    def _add_obsoletes(self):
+        global package_whitelist
+        final_query = self.base.sack.query()
+        if package_whitelist:
+        #  If anything obsoletes something that we have whitelisted ... then
+        # whitelist that too.
+            whitelist_query = self.base.sack.query().filter(name=package_whitelist)
+            obsoletes_query = self.base.sack.query().filterm(obsoletes=list(whitelist_query))
```

Comment:
why `filterM` here? According to docs, it is "Similar to [dnf.query.Query.filter()](https://dnf.readthedocs.io/en/latest/api_queries.html#dnf.query.Query.filter) but it modifies the query in place." but I don't see a reason to do this in place here?

Reply:
I would recommend using `filterm` on both places, it saves some copying. `.filter` returns a new copy of the query with filter applied, `filterm` avoids copying and applies the filter in place.